### PR TITLE
Typings improvments

### PR DESCRIPTION
**Pull request by aleek**
I added ConstraintTypes and Contraints definition. Nice to have, also I like the intellisense in VSC.

1. Tslint ran and all bugs fixed
2. unit tests run, 0 bugs.

Please review. If you like that, I'll add TypeScript Unit tests to jasmine.

### CODE REVIEW

**Comment by aleek** [comment]
Now we are ready!

**Comment by ElianCordoba** [comment]
Any updates on this?

**Comment by aleek** [comment]
@ElianCordoba what updates you'd like to see?

**Comment by aleek** [comment]
Development is IMO currently done. Depends if someone thinks, I can improve something. Feel free to dev if you'd like.

**Comment by drew-r** [comment]
I've had success with a mapped type for constraints.
e.g
`validate<T>(attributes: T, constraints: Partial<{ [P in keyof T]: any }>)`

**Comment by zepatrik** [comment]
Seems like @ansman won't merge it?

**Comment by grbspltt** [comment]
any chance of getting this PR merged?

File: validate.d.ts [hunk 1]
```diff
@@ -1,54 +1,323 @@
...
+}
+
+export interface ValidateJS {
+  validators: any;
+  formatters: any;
+
+  (attributes: any, constraints: Constraints, options?: ValidateOption): any;
+  validate(attributes: any, constraints: Constraints, options?: ValidateOption): any;
+  async(attributes: any, constraints: Constraints, options?: AsyncValidateOption): Promise<any>;
+  single(value: any, constraints: Constraints, options?: ValidateOption): any;
```
Comment (stramel) [inline validate.d.ts]: Shouldn't `constraints` on `single be of type `ConstraintsTypes` since it doesn't map to a key/name?

File: validate.d.ts [hunk 2]
```diff
@@ -1,54 +1,323 @@
...
+  numericality?: NumericalityConstraint;
+
+  presence?: PresenceConstraint;
+
+  url?: UrlConstraint;
+
+}
+
+export interface Constraints {
+  [name: string]: ConstraintsTypes;
```
Comment (stramel) [inline validate.d.ts]: NIT: Could we make use of `Record` here? `export type Constraints = Record<string, ConstraintsTypes>`

File: validate.d.ts [hunk 3]
```diff
@@ -1,54 +1,323 @@
...
+  nullify?: boolean;
+  trim?: boolean;
+}
+
+export interface ValidateJS {
+  validators: any;
+  formatters: any;
+
+  (attributes: any, constraints: Constraints, options?: ValidateOption): any;
+  validate(attributes: any, constraints: Constraints, options?: ValidateOption): any;
```
Comment (stramel) [inline validate.d.ts]: Can you improve the return types on these? Maybe something like, 
validate - `Record<string, string[]> | undefined>`
single - `string[] | undefined`
?

**Comment by sarangjo** [comment]
Any chance that this can be merged in? The `any` types are really frustrating to work with.

**Comment by zepatrik** [comment]
What about adding either this typing or #302's to @DefinitelyTyped ? Who would be up for a PR there @sarangjo @aleek @mauricedoepke ? I could do it but not now...

**Comment by sarangjo** [comment]
Will that override the typings provided by this repo?

**Comment by zepatrik** [comment]
Hm right, good point. After a quick search I would guess one could do a combination of https://github.com/microsoft/TypeScript/issues/25495 and https://stackoverflow.com/questions/40322788/how-to-overwrite-incorrect-typescript-type-definition-installed-via-types-packa
Basically declaring the module in a `@types` package and using the following `tsconfig.json`
```json
{
"compilerOptions": {
    "module": "validate.js",
    "noImplicitAny": true,
    "typeRoots": [
        "./node_modules/@types"
    ]
}
}
```
But this get's really hacky and will probably not be merged by @DefinitelyTyped

**TL;DR** the best way probably is to use one of the forks with the correct types as a dependency instead, as proposed in this thread: https://www.reddit.com/r/typescript/comments/cpe0u1/overriding_a_third_party_librarys_exported_types/

Maybe someone with more time can try what works...
